kit/ForKit.cpp

```diff
@@ -281,39 +281,56 @@ static void cleanupChildren()
     pid_t exitedChildPid;
     int status = 0;
     int segFaultCount = 0;
+    int killedCount = 0;
 
     // Reap quickly without doing slow cleanup so WSD can spawn more rapidly.
     while ((exitedChildPid = waitpid(-1, &status, WUNTRACED | WNOHANG)) > 0)
     {
         const auto it = childJails.find(exitedChildPid);
         if (it != childJails.end())
         {
-            if (WIFSIGNALED(status) && (WTERMSIG(status) == SIGSEGV ||
-                                        WTERMSIG(status) == SIGBUS ||
-                                        WTERMSIG(status) == SIGABRT))
+            if (WIFSIGNALED(status))
             {
-                ++segFaultCount;
-
-                std::string noteCrashFile(it->second + "/tmp/kit-crashed");
-                int noteCrashFD = open(noteCrashFile.c_str(), O_CREAT | O_TRUNC | O_WRONLY, S_IRUSR | S_IWUSR);
-                if (noteCrashFD < 0)
-                    LOG_ERR("Couldn't create file: " << noteCrashFile << " due to error: " << strerror(errno));
+                if (WTERMSIG(status) == SIGSEGV || WTERMSIG(status) == SIGBUS ||
+                    WTERMSIG(status) == SIGABRT)
+                {
+                    ++segFaultCount;
+                    LOG_INF("Child " << exitedChildPid << " has crashed, will remove its jail ["
+                                     << it->second << "].");
+
+                    std::string noteCrashFile(it->second + "/tmp/kit-crashed");
+                    int noteCrashFD = open(noteCrashFile.c_str(), O_CREAT | O_TRUNC | O_WRONLY,
+                                           S_IRUSR | S_IWUSR);
+                    if (noteCrashFD < 0)
+                        LOG_ERR("Couldn't create file: " << noteCrashFile
+                                                         << " due to error: " << strerror(errno));
+                    else
+                        close(noteCrashFD);
+                }
+                else if (WTERMSIG(status) == SIGKILL || WTERMSIG(status) == SIGTERM)
```

Comment:
I thought that OOM was detectable from some high-bits set in the status (?) - can we separate known OOM kills from other kills and count / report on that ?

Thanks!

Reply:
Unfortunately after more research, there is no specific signal.
The oom killer just sends SIGKILL (sometimes SIGTERM first).

There aren't many good sources.
https://www.kernel.org/doc/gorman/html/understand/understand016.html

There seem to be a OOMKilled signal specific to docker.

Regardless logging kill signals is still useful and we can expect OOM to be the main origin.

Reply:
So - lets detect OOM killers separately since that is/was what was asked; and some code pointers:

https://github.com/torvalds/linux/blob/master/mm/oom_kill.c#L947
	do_send_sig_info(SIGKILL, SEND_SIG_PRIV, victim, PIDTYPE_TGID);
https://github.com/torvalds/linux/blob/master/kernel/signal.c#L1135

Should have siginfo's si_code == SI_KERNEL as a source == OOM. cf. https://man7.org/linux/man-pages/man2/sigaction.2.html

We should also detect the docker OOM and glup that into it which would be with an exit with code 137 and add those two.

Of course counting other types of KILL is fine too - lets report 3 numbers crashes, kills, OOM-kills back in the same method =)

Thanks !

Reply:
Did you read the above Meven ?

Reply:
I did.

But https://linux.die.net/man/2/waitid had the solution.

I found `waitid` to have siginfo.

Latest commit uses it.
Haven't tested the oomKilled counter yet, need to tweak things with cgroup.

Reply:
docker OOM, `137` status is just 128 (means status is a signal) + 9 (the signal value, SIGKILL).

It is my understanding that docker leverage Linux cgroup limits to have the kernel handle the oom killing itself https://docs.docker.com/engine/containers/resource_constraints/

Reply:
Ah ok =) fair enough; I assumed there was some other signalling here.